yarn-project/simulator/src/avm/journal/nullifiers.ts

```diff
@@ -0,0 +1,156 @@
+import { Fr } from '@aztec/foundation/fields';
+
+import type { NullifiersDB } from '../../index.js';
+
+/**
+ * A class to manage new nullifier staging and existence checks during a contract call's AVM simulation.
+ * Maintains a pending nullifier cache, and ensures that existence checks fall back to the correct source.
+ * When a contract call completes, its pending nullifier set can be merged into its parent's.
+ */
+export class Nullifiers {
+  /** Cached nullifiers. */
+  private cache: PendingNullifiers;
```

Comment:
Similar to my comment for public storage, I'm open to merging the PendingNullifiers and Nullifiers classes. I like the separation, but it might not be necessary.

I'm also open to renaming suggestions for both!